internal/saucecloud/retry/junitretrier.go

```diff
@@ -17,80 +16,55 @@ type JunitRetrier struct {
 	VDCReader job.Reader
 }
 
-func getKeysFromMap(mp map[string]bool) []string {
-	keys := make([]string, len(mp))
-
-	i := 0
-	for k := range mp {
-		keys[i] = k
-		i++
-	}
-	return keys
-}
-
-func getFailedClasses(report junit.TestSuites) []string {
-	classes := map[string]bool{}
-
-	for _, s := range report.TestSuites {
-		for _, tc := range s.TestCases {
-			if tc.Error != "" || tc.Failure != "" {
-				classes[tc.ClassName] = true
-			}
-		}
-	}
-	return getKeysFromMap(classes)
-}
-
-func (b *JunitRetrier) retryOnlyFailedClasses(reader job.Reader, jobOpts chan<- job.StartOptions, opt job.StartOptions, previous job.Job) {
+func (b *JunitRetrier) retryFailedTests(reader job.Reader, jobOpts chan<- job.StartOptions, opt job.StartOptions, previous job.Job) {
 	content, err := reader.GetJobAssetFileContent(context.Background(), previous.ID, junit.JunitFileName, previous.IsRDC)
 	if err != nil {
-		log.Debug().Err(err).Msgf(msg.UnableToFetchFile, junit.JunitFileName)
+		log.Err(err).Msgf(msg.UnableToFetchFile, junit.JunitFileName)
 		log.Info().Msg(msg.SkippingSmartRetries)
 		jobOpts <- opt
 		return
 	}
 	suites, err := junit.Parse(content)
 	if err != nil {
-		log.Debug().Err(err).Msgf(msg.UnableToUnmarshallFile, junit.JunitFileName)
+		log.Err(err).Msgf(msg.UnableToUnmarshallFile, junit.JunitFileName)
 		log.Info().Msg(msg.SkippingSmartRetries)
 		jobOpts <- opt
 		return
 	}
 
-	classes := getFailedClasses(suites)
-	log.Info().
-		Str("suite", opt.DisplayName).
-		Str("attempt", fmt.Sprintf("%d of %d", opt.Attempt+1, opt.Retries+1)).
-		Msgf(msg.RetryWithClasses, strings.Join(classes, ","))
-
-	setClassesToRetry(&opt, classes)
-
+	setClassesToRetry(&opt, suites)
 	jobOpts <- opt
 }
 
 // setClassesToRetry sets the correct filtering flag when retrying.
 // RDC API does not provide different endpoints (or identical values) for Espresso
 // and XCUITest. Thus, we need set the classes at the correct position depending the
 // framework that is being executed.
-func setClassesToRetry(opt *job.StartOptions, classes []string) {
+func setClassesToRetry(opt *job.StartOptions, suites junit.TestSuites) {
+	lg := log.Info().
+		Str("suite", opt.DisplayName).
+		Str("attempt", fmt.Sprintf("%d of %d", opt.Attempt+1, opt.Retries+1))
+
 	if opt.Framework == xcuitest.Kind {
-		opt.TestsToRun = classes
+		opt.TestsToRun = junit.GetFailedTests(suites)
+		lg.Msgf(msg.RetryWithClasses, opt.TestsToRun)
 		return
 	}
 	if opt.TestOptions == nil {
 		opt.TestOptions = map[string]interface{}{}
 	}
-	opt.TestOptions["class"] = classes
+	opt.TestOptions["class"] = junit.GetFailedClasses(suites)
+	lg.Msgf(msg.RetryWithClasses, opt.TestOptions["class"].([]string))
```

Comment:
You are only casting because you erased the type beforehand. Just keep it:
```suggestion
        classes := junit.GetFailedClasses(suites)
	opt.TestOptions["class"] = classes
	lg.Msgf(msg.RetryWithClasses, classes)
```